Feature/aidumov/dp 5 onboarding screen

superaydumov opened the pull request:
PR в соответствии с выполненной задачей DP-5 по добавлению онбординга пользователя при первом запуске приложения.
После прохождения онбординга пользователь перекидывается на AuthFlow (AuthViewController).

superaydumov committed:
[DP-5] - Создал онбординг для пользователя, который показывается при первом запуске приложения (использовал UICollectionView и UIPageControl).
Создал класс CustomButtons с переиспользуемыми кнопками.
Обновил ассеты, добавил недостающие шрифты.

superaydumov committed:
Merge branch 'develop' into feature/aidumov/DP-5-Onboarding-screen

# Conflicts:
#	DavaiPosmotrim.xcodeproj/project.pbxproj
#	DavaiPosmotrim/SceneDelegate.swift

SKemenov commented on DavaiPosmotrim/OnboardingViewController/OnboardingCollectionViewCell.swift:
```diff
@@ -0,0 +1,129 @@
+
+final class OnboardingCollectionViewCell: UICollectionViewCell {
+
+    // MARK: - Stored proprties
+
+    static let reuseIdentifier = "OnboardingCollectionViewCell"
+    private var cellConstraints = [NSLayoutConstraint]()
+
+    // MARK: - Computed proprties
```
В марках опечатка -> надо properties

superaydumov replied:
Done

SKemenov commented on DavaiPosmotrim/OnboardingViewController/OnboardingCollectionViewCell.swift:
```diff
@@ -0,0 +1,129 @@
+        super.init(frame: frame)
+        self.contentView.backgroundColor = .clear
+
+        addSubviews()
+        constraintsSetup()
```
Нейминг функций не по договоренности (см. readme.md)

superaydumov replied:
Done

SKemenov commented on DavaiPosmotrim/OnboardingViewController/OnboardingCollectionViewCell.swift:
```diff
@@ -0,0 +1,129 @@
+        return label
+    }()
+
+    // MARK: - Lifecycle
```
`// MARK: - Lifecycle` следует использовать для методов жизненного цикла VC - `viewDidLoad`, `ViewWillAppear` и тд.
Рекомендую тут использовать `// MARK: - Inits`

superaydumov replied:
Done

SKemenov commented on DavaiPosmotrim/OnboardingViewController/OnboardingCollectionViewCell.swift:
```diff
@@ -0,0 +1,129 @@
+        [cellImageView,
+        cellLowerLabel,
+        cellUpperLabel
+        ].forEach {
+            contentView.addSubview($0)
+            $0.translatesAutoresizingMaskIntoConstraints = false
+        }
```
Классно сделал!)

SKemenov commented on DavaiPosmotrim/OnboardingViewController/OnboardingCollectionViewCell.swift:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    private func addImageViewConstraints(using indexPath: IndexPath) {
+        if indexPath.row == 1 {
+            cellConstraints += [
+                cellImageView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor),
+                cellImageView.trailingAnchor.constraint(equalTo: contentView.trailingAnchor)
+            ]
+        } else {
+            cellConstraints += [
+                cellImageView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 32),
+                cellImageView.trailingAnchor.constraint(equalTo: contentView.trailingAnchor, constant: -32)
+            ]
+        }
```
Как вариант тут можно сделать тернарный оператор, например для leading констрейнта
```` swift
cellImageView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: indexPath.row == 1 ? 0 : 32),

```

superaydumov replied:
Хорошая идея! Переписал с учетом комментария

SKemenov commented on DavaiPosmotrim/OnboardingViewController/OnboardingViewController.swift:
```diff
@@ -0,0 +1,310 @@
+        )
+    ]
+
+    // MARK: - Computed properties
```
Тут не совсем корректный марк. Это lazy view, они высчитываются один раз при первом обращении. Computed properties - каждый раз высчитывается при обращении к ним.

superaydumov replied:
Заменил на Lazy Properties. Хотя, честно говоря, особенной разницы не вижу
Потому что они всё равно высчитываются..хоть и один раз))

SKemenov commented on DavaiPosmotrim/OnboardingViewController/OnboardingViewController.swift:
```diff
@@ -0,0 +1,310 @@
+        layout.scrollDirection = .horizontal
+        let collectionView = UICollectionView(frame: .zero, collectionViewLayout: layout)
+        collectionView.backgroundColor = .clear
+        collectionView.register(
+            OnboardingCollectionViewCell.self,
+            forCellWithReuseIdentifier: OnboardingCollectionViewCell.reuseIdentifier)
```
Не по договоренностям (readme.md -> Перенос скобок)
Если параметры функции с новой строки, то закрывающую скобку рекомендуется перенести на новую строку.

superaydumov replied:
Done

SKemenov commented on DavaiPosmotrim/OnboardingViewController/OnboardingViewController.swift:
```diff
@@ -0,0 +1,310 @@
+        return pageControl
+    }()
+
+    // MARK: - Lifecycle
```
аналогично, тут лучше марк inits

superaydumov replied:
Done

SKemenov commented on DavaiPosmotrim/OnboardingViewController/OnboardingViewController.swift:
```diff
@@ -0,0 +1,310 @@
+        toogleButtons()
+    }
+
+    // MARK: - Private methods
```
Как вариант - все private methods можно вынести в `private extension OnboardingViewController { }`

SKemenov commented on DavaiPosmotrim/OnboardingViewController/OnboardingViewController.swift:
```diff
@@ -0,0 +1,310 @@
+
+        addSubviews()
+        constraintsSetup()
+        toogleButtons()
```
Нейминг функций не по правилам

superaydumov replied:
Поправил

SKemenov commented on DavaiPosmotrim/OnboardingViewController/OnboardingViewController.swift:
```diff
@@ -0,0 +1,310 @@
+        ])
+    }
+
+    private func toogleButtons() {
```
опечатка - нужно toggleButtons

superaydumov replied:
Done

SKemenov commented on DavaiPosmotrim/OnboardingViewController/OnboardingViewController.swift:
```diff
@@ -0,0 +1,310 @@
+extension OnboardingViewController: UICollectionViewDataSource {
+
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return pages.count
```
тут ключевое слово `return` можно опускать

SKemenov commented on DavaiPosmotrim/OnboardingViewController/OnboardingViewController.swift:
```diff
@@ -0,0 +1,310 @@
+    }
+}
+
+    // MARK: - OnboardingViewProtocol
+
+extension OnboardingViewController: OnboardingViewProtocol {
+
+}
```
неиспользуемый код. Рекомендую удалить. Протокол кстати тоже пустой. Какое его назначение?

superaydumov replied:
Насколько я понял, это протокол вьюконтроллера, с помощью которого он связывается с презентером для вызовы необходимых методов.
Добавил марк, чтобы не забыть добавить код, если понадобится.

SKemenov replied:
ок

SKemenov commented on DavaiPosmotrim/UI/CustomButtons.swift:
```diff
@@ -0,0 +1,45 @@
+
+import UIKit
+
+final class CustomButtons: UIView {
```
См. комментарии к предыдущим файлам

superaydumov replied:
Done

SKemenov reviewed (changes requested):
Молодец, мне понравился код! 
Есть мелкие недочеты, просто нужно привыкнуть к новым правилам)

superaydumov committed:
[DP-5] - Исправил замечания после комментариев:
- поправил названия методов для соответствия правилам;
- переименовал некоторые "MARK";
- отрефакторил метод setupImageViewConstraints в классе OnboardingCollectionViewCell.

IuriiKarpunov commented on DavaiPosmotrim/OnboardingViewController/OnboardingCollectionViewCell.swift:
```diff
@@ -0,0 +1,128 @@
+
+        return attrString
+    }
+
```
Лишний перенос строки

superaydumov replied:
Done

superaydumov committed:
[DP-5] - Исправил отступ в методе adjustCellUpperLabel

SKemenov reviewed (approved):
Супер!

uistelmakh commented on DavaiPosmotrim/OnboardingViewController/OnboardingViewController.swift:
```diff
@@ -0,0 +1,313 @@
+    }
+
+    @objc func didTapBeginButton(sender: AnyObject) {
+        UserDefaults.standard.setValue(true, forKey: "isOnboardingShown")
```
В константы вынести isOnboardingShown

superaydumov replied:
Done

uistelmakh commented on DavaiPosmotrim/OnboardingViewController/OnboardingCollectionViewCell.swift:
```diff
@@ -0,0 +1,127 @@
+
+    // MARK: - Stored proprties
+
+    static let reuseIdentifier = "OnboardingCollectionViewCell"
```
Создай Енам с константами

superaydumov replied:
Done
